### PR TITLE
Add horizontal Discovery use case

### PR DESCRIPTION
**Pull request by mmccool**
Addresses Issue https://github.com/w3c/wot-usecases/issues/6

### CODE REVIEW

**Comment by farshidtz** [comment]
@AndreaCimminoArriaga and I are writing requirements of [connected-building-energy-efficiency](https://github.com/w3c/wot-usecases/blob/master/USE-CASES/connected-building-energy-efficiency.md) and [smart-campus](https://github.com/w3c/wot-usecases/blob/master/USE-CASES/smart-campus.md) use cases as requested by @mlagally. Requirements are discovery related, such as:
- federated directories
- directory queries
- directory information model

[Draft PR](https://github.com/w3c/wot-architecture/pull/528)

I thought we would write those requirements with respect to the use cases, and those, along with other discovery requirements would be used as basis for defining discovery specs. 

I feel there is overlap between discovery as a use case and [requirements](https://github.com/w3c/wot-architecture/tree/master/REQUIREMENTS). Any suggestions?

**Comment by mlagally** [comment]
> @AndreaCimminoArriaga and I are writing requirements of [connected-building-energy-efficiency](https://github.com/w3c/wot-usecases/blob/master/USE-CASES/connected-building-energy-efficiency.md) and [smart-campus](https://github.com/w3c/wot-usecases/blob/master/USE-CASES/smart-campus.md) use cases as requested by @mlagally. Requirements are discovery related, such as:
> 
> * federated directories
> * directory queries
> * directory information model
> 
> [Draft PR](https://github.com/w3c/wot-architecture/pull/528)
> 
> I thought we would write those requirements with respect to the use cases, and those, along with other discovery requirements would be used as basis for defining discovery specs.
> 
> I feel there is overlap between discovery as a use case and [requirements](https://github.com/w3c/wot-architecture/tree/master/REQUIREMENTS). Any suggestions?

I don't think that overlaps wrt requirements are a problem but rather a logical consequence of having system components that are used across use cases. 
If we have a requirement document addressing the needs of one use case is can be extended by additional input / requests from another UC.

**Comment by farshidtz** [comment]
> I don't think that overlaps wrt requirements are a problem but rather a logical consequence of having system components that are used across use cases.
> If we have a requirement document addressing the needs of one use case is can be extended by additional input / requests from another UC.
True, but my concern was related to overlaps between "discovery use case" and discovery requirements derived from other business/system use cases. I wasn't fully involved in the discussions; is there any document I can refer to, to understand the scope of use cases?

**Comment by mmccool** [comment]
We definitely need to connect the horizontal and vertical use cases somehow.  I was planning to document how some of the requirements are driven by particular verticals.  For example, "geospatial query" is not needed for home discovery, but is useful for smart city discovery.

**Comment by jenlincode** [comment]
(As mentioned in the Use Case call:) The only clarification I was curious about while reviewing this was what the scope of the discovery network. Is this to be local, cross-network, etc? Although it is totally understandable if the intent is for the scope to be generic or undefined.

**Comment by mmccool** [comment]
Also discussed in Discovery call, there might be cases where the identity of a querier should not be revealed to the Directory service (e.g. they might get a token from an auth server, then present that token to the directory).  However this causes problems if we want to control access to particular *dynamic* resources, eg. registered TDs.   We would need per-registered-TD scopes and there is no easy way to do that in TDs (if we use static TDs.  It is possible if we use dynamic TDs, but then we either have to list everything, causing a privacy leak, or filter it for the user, which means we have to identify the user, bringing us back to the same problem...)

**Comment by mmccool** [comment]
We should also consider end-to-end security, eg. JOSE.   Do we have a use case where we need this for metadata, eg. TDs?   We will have TLS but if the connection gets "terminated" at an intermediary (eg a proxy) there is some chance the proxy could "spy" on the content.  Object security would add an extra layer of protection.   Security schemes for object security are not defined yet though so we can't add them to the TD for directories (except only as a contentType) and be conformant with the 1.x spec.

**Comment by mmccool** [comment]
but object security (eg JOSE) *could* be an additional requirement under "Privacy Considerations"

**Comment by mmccool** [comment]
Another requirement is that we need local, "offline" security.  Right now https depends on checking for revoked certificates, which requires external internet access...

File: USE-CASES/discovery.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+Discovery is the general process whereby WoT Thing Descriptions satisfying a specific set of such
+constraints are retreived by a running system.
+
+#### Variants:
+
+* Run-time discovery allows late binding of orchestration services to particular devices and requires that
+consumers be able to adapt to Thing Descriptions discovered when a service is deployed.
+* Development-time discovery may be useful during system development to build services that can interface to
+a particular class of Thing Descriptions.  In this case what actually needs to be discovered Thing Description
+Templates, not specific Thing Descriptions.
```
Comment (mlagally) [inline USE-CASES/discovery.md:79]: Template -> Thing Models

File: USE-CASES/discovery.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+### Gaps:
+
+* The current WoT standards define a metadata format (the Thing Description) but not a means of distributing it.
+
+### Existing standards:
+
+* WoT Thing Description
+* CoreRD
+* DID
```
Comment (mlagally) [inline USE-CASES/discovery.md:104]: Links please

**Comment by mlagally** [comment]
Reviewed in use-case call on 17.9.

**Pull request merged**